src/v/utils/vint.h

```diff
@@ -12,12 +12,36 @@
 #pragma once
 #include "bytes/bytes.h"
 
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/iostream.hh>
+
 #include <cstdint>
 
 namespace unsigned_vint {
 /// At most 5 bytes are needed to encode a 32 bit value
 inline constexpr size_t max_length = 5;
 
+inline ss::future<std::pair<uint32_t, size_t>>
+deserialize_via_stream(ss::input_stream<char>& s) {
```

Comment:
There's a lot of code duplication between this method and the one below it called `deserialize`. However this method is different in the fact that the data is obtained by reading exactly 1 byte at a time, while `deserialize` operates on a range of bytes. Ideas on how to genericise the algorithm portion would be appreciated here.

Reply:
Interesting, I thought you could do it with a helper taking an iterator concept, with a boolean template argument to make it a coroutine or not (since your new method is a coroutine but the existing de-serialize isn't), like so:

(shows the common helper method `_impl` and the old call calling into it along with the adaptor for the `Range` type)

```
template <typename R>
struct RangeReader {

    using itr_type = decltype(std::begin(std::declval<R&>()));

    explicit RangeReader(R &&r) : _i(std::begin(r)), _e(std::end(r)) {}

    bool more() {
        return _i != _e;
    }

    auto next() {
        auto r = *_i;
        ++_i;
        return r;
    }

    itr_type _i, _e;

};

template<bool coro, typename Reader>
inline std::pair<uint64_t, size_t>
deserialize_impl(Reader& r, uint8_t limit) noexcept {
    uint64_t result = 0;
    size_t bytes_read = 0;
    uint64_t shift = 0;
    while (shift <= limit && r.more()) {
        uint64_t byte;
        if constexpr (coro) {
            byte = co_await r.next();
        } else {
            byte = r.next();
        }
        bytes_read++;
        if (byte & 128u) {
            result |= ((byte & 127u) << shift);
        } else {
            result |= byte << shift;
            break;
        }
        shift += 7;
    }

    auto ret = std::make_pair(result, bytes_read);
    if constexpr (coro) {
        co_return ret;
    } else {
        return ret;
    }
}


template<typename Range>
inline std::pair<uint64_t, size_t>
deserialize(Range&& r, uint8_t limit) noexcept {
    auto&& reader = RangeReader<Range>{r};
    return deserialize_impl<false>(reader, limit);
}
```

However, it doesn't work: you can't make a method "conditionally" a coroutine even with `if constexpr`, some justification [here](https://github.com/GorNishanov/coroutines-ts/issues/29) and standard language in [this answer](https://stackoverflow.com/a/71433344).

So I guess you could still factor out most of the logic into a helper function but you'll still need a skeleton function for each impl that contains the loop (since the `co_await` is in there) and the `co_return`.

Well, you could still use the single impl that I show, but it would have to be a coroutine, and the old synchronous method could call into it and unwrap the future immediately, but that just seems nasty (and possibly very slow depending on how much the compiler is able to see through those tricks).

Reply:
Here's how I'd do it instead: with a decoder helper struct which keeps all the state, so you only need the minimal DRY skeleton functions to expose the `co_await` and `co_return`:

```
namespace unsigned_vint {

namespace detail {

template<typename ResultType>
struct var_decoder {
    ResultType result = 0;
    size_t bytes_read = 0;
    uint64_t shift = 0;
    uint8_t limit;

    explicit var_decoder(uint8_t limit)
      : limit(limit) {}

    /**
     * @brief Incorporate a new byte into the state.
     * Returns true if the outer loop should break (last byte found).
     */
    bool accept(uint64_t byte) {
        bytes_read++;
        if (byte & 128) {
            result |= ((byte & 127) << shift);
        } else {
            result |= byte << shift;
            return true;
        }
        shift += 7;
        return shift > limit;
    }

    std::pair<ResultType, size_t> get_result() { return {result, bytes_read}; }
};

} // namespace detail

/// At most 5 bytes are needed to encode a 32 bit value
inline constexpr size_t max_length = 5;

inline ss::future<std::pair<uint32_t, size_t>>
deserialize_via_stream(ss::input_stream<char>& s) {
    detail::var_decoder<uint32_t> decoder(8 * sizeof(uint32_t));
    while (!s.eof()) {
        if (decoder.accept((co_await s.read_exactly(1))[0])) {
            break;
        }
    }
    co_return decoder.get_result();
}

namespace detail {

/// \brief almost identical impl to leveldb, made generic for c++
/// friendliness
/// https://github.com/google/leveldb/blob/201f52201f/util/coding.cc#L116
template<typename Range>
inline std::pair<uint64_t, size_t>
deserialize(Range&& r, uint8_t limit) noexcept {
    var_decoder<uint64_t> decoder(limit);
    for (auto byte : r) {
        if (decoder.accept(byte)) {
            break;
        }
    }
    return decoder.get_result();
}
} // namespace detail
```

Reply:
This I think would work really really nicely. Thanks for the advice!